### PR TITLE
FE-1317 - Flash component

### PR DESCRIPTION
**Pull request by dawidzarzycki**
# Description
- Create `Flash` component with the DLS standard
- Extend props in the `Toast` component to be reusable with the `Flash` component
- Fix warnings in `storybook`

# TODO
- Unit Tests

# Screenshots / Gifs
- Classic
Nagranie-z-ekranu-2019-06-4-o-13 26 26
- DLS
Nagranie-z-ekranu-2019-06-4-o-13 53 22

# Testing Instructions
- Toast component -  QA needed

### CODE REVIEW

File: src/components/flash/flash-legacy.component.js [hunk 1]
```diff
@@ -344,89 +221,54 @@ class Flash extends React.Component {
...
+          { this.formatDescription(this.description) }
+        </FlashMessageStyle>
+        {!this.props.timeout && (
+          <Icon
+            className='carbon-flash__close'
+            data-element='close'
+            key='close'
+            onClick={ this.props.onDismiss }
+            type='close'
+          />
```
Comment (grabkowski) [inline src/components/flash/flash-legacy.component.js]: it could be a styled component

File: src/components/flash/flash-legacy.spec.js [hunk 2]
```diff
@@ -1,46 +1,66 @@
 import React from 'react';
 import TestUtils from 'react-dom/test-utils';
-import Flash from './flash';
 import { shallow, mount } from 'enzyme';
-import { elementsTagTest, rootTagTest } from '../../utils/helpers/tags/tags-specs';
-import Portal from './../portal';
-import guid from '../../utils/helpers/guid';
+import FlashLegacy from './flash-legacy.component';
```
Comment (grabkowski) [inline src/components/flash/flash-legacy.spec.js]: you could just import `Flash` and leave everything as it is. It's default export, so the name does not matters.

File: src/components/flash/flash-legacy.spec.js [hunk 3]
```diff
@@ -351,57 +390,32 @@ describe('Flash', () => {
...
+      const iconVariant = {
+        default: 'default',
+        success: 'tick',
+        maintenance: 'settings',
+        info: 'info',
+        help: 'question',
+        error: 'error',
+        new: 'gift',
+        warning: 'warning'
+      };
```
Comment (grabkowski) [inline src/components/flash/flash-legacy.spec.js]: that could be placed directly in `.each` method

File: src/components/flash/flash.scss [hunk 4]
```diff
@@ -103,65 +51,4 @@
...
-.carbon-flash {
-  &--error, &--new, &--info {
-    .carbon-flash__message,
-    .carbon-flash__close,
-    .carbon-flash__icon:before {
-      color: white;
-    }
-  }
-}
+}
```
Comment (grabkowski) [inline src/components/flash/flash.scss]: do we need this file?

File: src/components/flash/flash.style.js [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+const colors = {
+  warning: '#FF7D00',
+  default: '#335B6D',
+  error: '#c7384f',
+  info: '#1573E6',
+  new: '#663399',
+  success: '#50B848',
+  help: '#FFAB00',
+  maintenance: '#FF7D00'
+};
```
Comment (grabkowski) [inline src/components/flash/flash.style.js]: shouldn't these colors be defined in a theme config?

File: src/components/flash/flash-legacy.component.js (hunk 1 above)
Comment (dawidzarzycki) [reply src/components/flash/flash-legacy.component.js]: yup, done

File: src/components/flash/flash-legacy.component.js [hunk 6]
```diff
@@ -344,89 +221,54 @@ class Flash extends React.Component {
...
-      contents.push(
+    return (
+      <FlashContentStyle>
         <Icon
-          className='carbon-flash__close'
-          data-element='close'
-          key='close'
-          onClick={ this.props.onDismiss }
-          type='close'
+          className='carbon-flash__icon' type={ this.iconType }
```
[the remark this answers is not in the document]
Comment (dawidzarzycki) [reply src/components/flash/flash-legacy.component.js]: after moved to styled-components we don't need, you are right :)

File: src/components/flash/flash-legacy.spec.js (hunk 3 above)
Comment (dawidzarzycki) [reply src/components/flash/flash-legacy.spec.js]: fixed

File: src/components/flash/flash-legacy.spec.js (hunk 2 above)
Comment (dawidzarzycki) [reply src/components/flash/flash-legacy.spec.js]: I used FlashLegacy name to marked that this is Legacy component

File: src/components/flash/flash.scss (hunk 4 above)
Comment (dawidzarzycki) [reply src/components/flash/flash.scss]: not anymore :)

File: src/components/flash/flash.style.js (hunk 5 above)
Comment (dawidzarzycki) [reply src/components/flash/flash.style.js]: This is classic style

File: src/components/flash/flash-legacy.style.js [hunk 7]
```diff
@@ -0,0 +1,193 @@
+import styled, { css } from 'styled-components';
+import Link from '../link';
+import Icon from '../icon/icon';
```
Comment (izakrzewska) [inline src/components/flash/flash-legacy.style.js]: double icon

File: src/components/flash/flash-legacy.component.js [hunk 8]
```diff
@@ -456,4 +293,32 @@ class Flash extends React.Component {
...
+  /** Time for flash to remain on screen */
+  timeout: PropTypes.oneOfType([
+    PropTypes.string,
+    PropTypes.number
+  ])
+};
+
+FlashLegacy.defaultProps = {
+  as: 'success',
+  className: '',
```
[the remark this answers is not in the document]
Comment (dawidzarzycki) [reply src/components/flash/flash-legacy.component.js]: nope, fixed

**Comment by harpalsingh** [comment]
The new component has some issues that need to be reviewed. The animation itself is clunky and seems to be due to the way it is rendered and animated.

This is what we see currently:

flash-howitlooksnow

This is how the animation should flow:

flash-how-it-should-look

As you can see the 2nd visual is a lot smoother, you can view it at http://designsystem.sage.com/foundations/colour/ simply click on colour to copy. While the general animation start and end mimic the requirement, due to the way it is implemented it has a very clunky feel and animation to it, can this pls be fixed, after which we can re-review.

**Comment by harpalsingh** [comment]
I've been updating the code style to reflect the animation intention in the original design, also fixed the center alignment of the icons in the component.

flash

File: src/components/toast/toast.style.js [hunk 9]
```diff
@@ -12,31 +12,30 @@ import { THEMES } from '../../style/themes';
...
+  right: ${({ isCenter }) => (isCenter ? '0px' : '30px')};
   top: 0;
-  transform:  ${({ isCenter }) => (isCenter ? 'translateX(50%)' : '')}; 
 
   ${({ theme }) => theme.name !== THEMES.classic && css`
     &${animationName}-appear,
     &${animationName}-enter {
       opacity: 0;
-      transform: ${({ isCenter }) => (isCenter ? 'translateX(50%) scale(0.5)' : 'scale(0.5)')};
+      transform: ${({ isCenter }) => (isCenter ? ' scale(0.5)' : 'scale(0.5)')};
```
Comment (izakrzewska) [inline src/components/toast/toast.style.js]: does it change anything?

Comment (dawidzarzycki) [reply src/components/toast/toast.style.js]: nope XD good point!

File: src/components/toast/__snapshots__/toast.spec.js.snap [hunk 10]
```diff
@@ -17,34 +18,77 @@ exports[`Toast when toast is closed renders null 1`] = `
...
+                "displayName": "styled.div",
+                "foldedComponentIds": Array [],
+                "render": [Function],
+                "styledComponentId": "sc-bZQynM",
+                "target": "div",
+                "toString": [Function],
+                "warnTooManyClasses": [Function],
+                "withComponent": [Function],
+              }
+            }
```
Comment (grabkowski) [inline src/components/toast/__snapshots__/toast.spec.js.snap]: that shouldn't be included in a snapshot

**Pull request merged**